### DISCUSSION TITLE
Single component slint found, multiple components not found

### DISCUSSION THREAD

**Discussion by norricorp** [slint-ui/slint#4695] [Q&A]
First many thanks for the several answers to my previous question.
I am continuing with this code, for the moment with the slint! macro approach but will also do the file.slint approach asfter.

So this works
```
 use slint::slint;

fn main() {

    let main_window = C1::new().unwrap();
    // etc etc 
 }

slint! {
	import { Button, VerticalBox, LineEdit, StandardButton } from "std-widgets.slint";
	
	export component C1 inherits Window {
// etc etc 
       }
}
```

C1 is found, it compiles, happiness ensues. But ...
```
 use slint::slint;

fn main() {

    let main_window = C1::new().unwrap();
    // etc etc 
 }

slint! {
	import { Button, VerticalBox, LineEdit, StandardButton } from "std-widgets.slint";
	
	export component C1 inherits Window {
// etc etc 
       }

       export component C2 inherits Dialog {
// etc etc 
       }
}
```
There is now a compile error saying use of undeclared type C1. Later in the code, I use C2 and that is found. From what I can see multiple components can be exported. 
I realise I am not showing the full code but according to VSC my braces are matching. Can there be a Window and Dialog in the same slint macro?
Any other ideas?

regards,
John

**Comment by ogoffart** [comment]
This is not yet supported (tracked in https://github.com/slint-ui/slint/issues/784 )
Several slint! macro works, but they can't share the same structs.

**Comment by norricorp** [comment]
So I am better off going down the file.slint route?

Interesting - just made changes to the slint file approach and get exactly same error "use of undeclared type `C1`"

Just looked at issue 784. So do I need multiple .slint files? Or can only one component be exported for the whole application? (ie the last exported component will be exported)

**Comment by norricorp** [comment]
I am still not sure what the rules are here.

The suggestion is that I can have multiple slint! macros but having tried that, C1 component is still undefined

And the same is true of using a slint file; that can only have one exported component.

The todo example on master branch has the Dialog inside a popupWindow inside an exported component so only a single exported component. But the code on branch release1.4 is very different. Is the code on master the latest, greatest and approved? (code on other branches may be later and greater but if not approved then irrelevant)

So can I use multiple slint files and change build.rs to have a list of files? eg
      slint_build::compile("ui/c1.slint", "ui/c2.slint").unwrap();

**Comment by norricorp** [comment]
Just to add. 
Changed build.rs to 
fn main() {
    slint_build::compile("ui/c1.slint").unwrap();
    slint_build::compile("ui/c2.slint").unwrap();
}
and that does compile
BUT
Having split the original slint file into holding C1 and C2, main.rs still shows 

let ui = C1::new()?;
  |              ^^^^ use of undeclared type `C1`
  
  Am I trying to the impossible?

**Comment by ffred** [comment]
Oops, just came with the same "problem"...
I probably should learn a little more of Slint to see what is the way to do things.

**Comment by vonnieda** [comment]
This quote from the issue is the important bit:

> Currently, only the last component in the main .60 (.slint) file that is compiled is "exported". In Rust a struct and in C++ a class is generated for it, with the known show(), etc. functions and accessors for declared properties. It also provides API to access singletons.

Basically, you have to have everything funnel up to a single component - your main app component. You can only have instances of that one component, and there is no way to reference the other components. You interact by changing state at that top level and the lower level components can access that.

I've found this incredibly confusing, and quite different from other UI frameworks I've used. I personally think this is "broken", and forces a lot of coupling between components, but it is possible to work around. 

Where I have landed is to have a top level components, with "models" for each sub-component. Callbacks go into a global singleton and then I dole those out within the Rust code.

Here is the start of my main app window, for example:

```
export component AppWindow inherits Window {
    in-out property <Page> page: Settings;

    // TODO Note to self: Currently you can't expose properties on
    // sub-components, so we have to bubble them all to the top.
    // See: https://github.com/slint-ui/slint/issues/784
    // Wait, how the heck do I hookup callbacks then?
    // Same way :(
    // Recent reading (find links) makes me think this should all move
    // to a global called, like AppLogic. I've started doing some of that
    // in AppState for callbacks.
    in-out property <HomeAdapter> home;
    in-out property <CardGridAdapter> search;
    in-out property <CardGridAdapter> artist-list;
    in-out property <ArtistDetailsAdapter> artist-details;
    in-out property <CardGridAdapter> release-group-list;
    in-out property <ReleaseGroupDetailsAdapter> release-group-details;
    in-out property <CardGridAdapter> release-list;
    in-out property <ReleaseDetailsAdapter> release-details;
    in-out property <CardGridAdapter> genre-list;
    in-out property <GenreDetailsAdapter> genre-details;
    in-out property <TrackListAdapter> track-list;
    in-out property <TrackDetailsAdapter> track-details;
    in-out property <CardGridAdapter> playlist-list;
    in-out property <SettingsAdapter> settings;
    in-out property <PlayerBarAdapter> player-bar;

```

Jason
